### ISSUE TITLE
Infinite Loop during getRecord

### ISSUE THREAD

**Issue by guibou** [Gabriel439/Haskell-Optparse-Generic-Library#9]
This short program runs until out of memory:

``` haskell
{-# LANGUAGE DeriveGeneric     #-}
{-# LANGUAGE OverloadedStrings #-}

import GHC.Generics
import Options.Generic

data Cmds = A | B | C
  deriving (Generic)

instance ParseRecord Cmds

main = (getRecord "This Fails") :: IO Cmds
```

Note: removing one constructor of the sum gives the excepted output. Adding one (e.g. 4 constructors) results in the following compilation error on the `instance ParseRecord Cmds` line:

```
OptParse.hs:10:10:
    No instance for (Options.Generic.GenericParseRecord
                       ((C1 Main.C1_0Cmds U1 :+: C1 Main.C1_1Cmds U1)
                        :+: (C1 Main.C1_2Cmds U1 :+: C1 Main.C1_3Cmds U1)))
      arising from a use of ‘Options.Generic.$gdmparseRecord’
    In the expression: Options.Generic.$gdmparseRecord
    In an equation for ‘parseRecord’:
        parseRecord = Options.Generic.$gdmparseRecord
    In the instance declaration for ‘ParseRecord Cmds’
```

(Edit: using stack with `lts-5.5` and ghc `7.10.3` on linux)

**Comment by Gabriella439** [comment]
Thanks for reporting this!  This is a major bug.

I will look into this as soon as possible (hopefully tonight)

**Comment by Gabriella439** [comment]
Okay, so I fixed the first issue (hanging when parsing three elements) in 29365ac555662d15327a5d95efef5250459637dd.  However, the issue of failing on four or more alternatives still remains and I'm working on that righ tnow

**Comment by Gabriella439** [comment]
Alright, I also fixed the issue of parsing 4 or more alternatives in c6e845cbe77ff727fe0397c684e1c85906bb1f52.

Here's a post-mortem explaining what went wrong for those who are curious.  The correct version should have four instances:

``` haskell
-- Instance #1
instance
        (GenericParseRecord (f :+: g), GenericParseRecord (h :+: i))
    =>  GenericParseRecord ((f :+: g) :+: (h :+: i))

-- Instance #2
instance
        (Constructor c, GenericParseRecord f, GenericParseRecord (g :+: h))
    =>  GenericParseRecord (M1 C c f :+: (g :+: h))

-- Instance #3
instance
        (Constructor c, GenericParseRecord (f :+: g), GenericParseRecord h)
    =>  GenericParseRecord ((f :+: g) :+: M1 C c h)

-- Instance #4
instance
        (Constructor c1, Constructor c2, GenericParseRecord f1, GenericParseRecord f2)
    =>  GenericParseRecord (M1 C c1 f1 :+: M1 C c2 f2) where
```

The first problem was that instances 2 and 3 had an infinite loop, and they get triggered when you have three or more constructors.  That's why you got an infinite loop when you had three alternatives

The second problem was that instance 1 was missing and was necessary for data types with 4+ alternatives.  That is why the code failed to type-check for 4 or more alternatives.

I will go ahead and upload this bug fix to Hackage as `optparse-generic-1.0.1`

**Comment by Gabriella439** [comment]
Alright, the bug fix is up on Hackage as `optparse-generic-1.0.1` so I will mark this resolved.

**Comment by Gabriella439** [comment]
I also deprecated version `1.0.0` on Hackage
